### ISSUE TITLE
Incorporate hook mechanism from smart importer

### ISSUE THREAD

**Issue by tbm** [beancount/beancount#458]
smart_importer has an apply hook mechanism which allows you to apply functions to all entries during import.

I found this very useful in another project I was working on (a rules-based importer) and I now also use it to apply a function I wrote that contains business logic for the organization.

Since there's nothing smart_importer-specific about the hook mechanism, could this be moved into beancount itself?

Copying @tarioch from smart_importer.

**Comment by blais** [comment]
Hi Martin, Can you be more specific about the hook?
I'm not sure if I provided an example, but there's a way to run the importer as a one-liner function call providing the list of importer objects.

**Comment by tbm** [comment]
Yes, I use the hook like this:

```
from smart_importer import apply_hooks

from rules_importer import ApplyRules
from spi.importer import SPIImporter

CONFIG = [
    apply_hooks(ofx.Importer('....', 'Assets:BoA:Advantage-Checking'), [ApplyRules(config), SPIImporter()]),
...
```
SPIImporter() does:

```
class SPIImporter(ImporterHook):
    def __call__(self, importer, file, imported_entries, existing_entries):
        self.account = importer.file_account(file)
        return [
            categorize_entry(entry, self.account) if isinstance(
                entry, data.Transaction) else entry
            for entry in imported_entries
        ]
```

and `categorize_entry` contains the business logic for the organization (looking up A/R and A/P, etc).

The relevant code for the hook mechanism is here:
https://github.com/beancount/smart_importer/blob/master/smart_importer/hooks.py

The nice thing is that this can be applied to any importer regardless of its type or Python class.

**Comment by tarioch** [comment]
Also adding @yagebu who initially created the hooks.

**Comment by blais** [comment]
The thing I was going to refer to is this:
https://github.com/beancount/beancount/blob/master/beancount/ingest/scripts_utils.py#L260
Basically you can make your importer config into a Python program and call one function.
I should provide an example of this an document it (it took a number of pirouettes to get that working along with the .import file; you can see what I mean if you look at the top of that file. In a rewrite I'd probably get rid of this as well with the bean-(file|identify|extract) tools and just provide a library function and you'd always just write a scrip[t).

What you call a "hook" prior in this thread isn't really a hook, it's an adapter that wraps functionality that's distinct from the importers (a hook is a callback invoked from library code). I now realize that it doesn't provide what you need.
Well maybe: the function version of the importers was created so that you could provide a function to detect duplicates, here:
https://github.com/beancount/beancount/blob/master/beancount/ingest/scripts_utils.py#L60

It's really applied as a general files for the entries just before printing them in extract, here:
https://github.com/beancount/beancount/blob/master/beancount/ingest/extract.py#L197
You can "abuse" this by making it do something else, e.g.:
script_utils.ingest(config, my_function_that_processes_extracted_entries).

I can rename the parameter to a list of functions that get applied in sequence, and if they're not set, keep find_dupilicate_entries() as the default. I think this is a more elegant solutions than writing an adapter.

I thought I could also break this function apart in two functions, so you could (1) get the extracted entries and (2) invoke the printer, yourself,  but actually, that ingest() function invokes any of the three workflows so maybe that's not the best way.

Let me know what you think

**Comment by blais** [comment]
Fixed by ba7b4699d136ab02859013331d98b3c5971c841f
I also added an example in 9585b4572136ca6d3561fef076d04ca04dd457f2

**Comment by tbm** [comment]
Thanks, I took a brief look and will take a closer look later. I think that would work.

However, the mechanism from smart_importer has one advantage: you can apply different functions to different importers (or to different banks using the same importer).  This is useful in my case since I want to apply my business logic function to the bank importers but not to the other importers which already have all the logic they need built in.

**Comment by blais** [comment]
Sure. Send me a PR (please write a simple unit test).

**Comment by dnicolodi** [comment]
My solution for doing this is using class inheritance and a method decorator (from memory, some details may be off):

```
class Importer(importer.ImporterProtocol):
    def extract(self, file, existing_entries=None):
       ...

def smart(f):
    @wraps(f)
    def wrapped(self, *args, **kwargs):
        entries = f(self, *args, **kwargs):
        return classify(entries)
    return wrapped

class SmartImporter(Importer):
    @smart 
    def extract(self, *args, **kwargs):
        super().extract(*args, **kwargs)
```

where the `Importer` class is the standard importer definition, the `smart` decorator and the `classify` function come from my machine learning transactions classifier support code and the `SmartImporter` class is defined in the beancount importer script.

**Comment by bratekarate** [comment]
Is there any chance the `apply_hooks`  function could be included with beancount (maybe under another name, as was it was pointed out that `hooks` may not be the best wording)? The solution with `script_utils.ingest()` hooks does not work with fava (https://github.com/beancount/fava/issues/1184#issuecomment-731831291). Exposing the `apply_hooks` function and the `ImporterHook` interface through the beancount library would make it much easier to write custom hooks/adapters/postprocessors to decorate importers with and to chain them in a config. Are PRs welcome in that regard?

**Comment by dnicolodi** [comment]
Ingest code has been moved from Beancount to Beangulp https://github.com/beancount/beangulp. The way of defining the ingest script used by Fava has been removed: the ingest script is responsible to call the library entry point. I don't know how Fava plans to adapt to those changes.

Beangulp already has an hook interface, I don't think we want to have two slightly different ways of doing the same thing. I had a quick look at the `apply_hooks` function and I find it a convoluted way to wrap a function. Written in a more pythonic way it is a ~5 lines function that I don't think we have to carry in the Beangulp codebase.

**Comment by bratekarate** [comment]
I see the issue is more complex than I anticipated. The way beangulp and beancount handle hooks seems great to me, but unfortonately it does not allow to leverage the power of smart importer in a way I intended to. It is crucial to me to allow hooks to be executed *before* the smart importer hooks, as I want the smart importer hooks to have the least weight. Other hooks should filter the entries before smart importer does some magic on entries where any other filters with "hard" rules have not been applied. 

Sadly, as long as smart importer does not implement the beangulp hook interface, any hooks that need to be run before smart importer must implement the `ImporterHook` interface of smart importer and be applied with the `apply_hooks` method. And then there's the issue of fava handling the ingest hooks....oh well. I'll probably just have to live with the situation of depending on `apply_hooks` for now and hope everything will magically fit my use case at some point in the future :)

**Comment by dnicolodi** [comment]
I find `smart_importer` extremely over-engineered. You can implement the same in about 30 lines of easy Python code, thus if you don't like its interface you can simply do not use it and roll your own solution. I don't think anything on the Beancount/Beangulp side will adapt to accommodate the `smart_importer` interface.

**Comment by dnicolodi** [comment]
Thinking about it for a minute, I don't think that if you want to have some code to operate on the extracted entries before `smart_importer` you need to hook it up in the same way as `smart_importer`. If you have an importer:
```python
class Importer(ImporterProtocol):
    ...
    def extract(self, file, existing_entries):
        pass
    ...
```
you can simply subclass it and add any processing you want:
```python
class ExtraImporter(Importer):
    def extract(self, file, existing_entries):
        entries = super().extract(file, existing_entries)
        # do your processing here
        return entries
```
You can encapsulate and chain this nicely:
```python
class Super:
    def extract(self, file, existing_entries):
        entries = super().extract(file, existing_entries)
        # do super processing here
        return entries

class Duper:
    def extract(self, file, existing_entries):
        entries = super().extract(file, existing_entries)
        # do duper processing here
        return entries

class SuperDuperImporter(Super, Duper, Importer):
    pass

importer = SuperDuperImporter()
```

Edit: In the original posting I forgot a few `self` parameters to the class methods. That's what you get for typing code directly in Github issue comments...

**Comment by bratekarate** [comment]
Thanks for the input, but let's agree to disagree. It doesn't really matter if something is overengineered if it works great and is easy to interface with. The time spent developing it may be not worth it, but now it is there and works. 

I'm also not saying beangulp should cater for smart importer, I rather came to the conclusion that smart importer should adapt to the beangulp hook interface if possible in the future. Or, someone rewrites it in like 30 lines of "more pythonic" code :) 

About your inheritance example, that is precisely what I want to avoid. I think chaining hooks/applying filter like a pipeline is much easier, cleaner and more reproducible on multiple systems.  As I said, it is what it is, I can live with the status quo for now.

**Comment by dnicolodi** [comment]
> I think chaining hooks/applying filter like a pipeline is much easier, cleaner and more reproducible on multiple systems.

Can you please elaborate what you mean? The code I propose implements exactly a "filter like pipeline", just the syntax to assemble the pipeline is different, my example using Python language paradigms instead that inventing a plugin system. In particular I don't understand how it may fail to be "reproducible on multiple systems".

As for re-implementing `smart_importer` in a simpler and easier to understand way:

```python
def target(t):
    # approximation: the account we want to predict is not a Checking account or Cash
    accounts = []
    for p in t.postings:
        if p.units is None:
            continue
        parts = p.account.split(':')
        if 'Checking' in parts or 'Cash' in parts:
            continue
        accounts.append((p.units, p.account))
    accounts.sort(key=lambda x: x[0], reverse=True)
    return accounts[0][1]

class Smart:
    def extract(self, file, existing_entries):
        entries = super().extract(file, existing_entries)
        if not existing_entries:
            return entries

        narrations = [ t.payee or t.narration for t in data.filter_txns(existing_entries) ]
        targets = [ target(t) for t in data.filter_txns(existing_entries) ]
        vectorizer = TfidfVectorizer()
        X = vectorizer.fit_transform(narrations)
        classifier = LinearSVC().fit(X, targets)
        
        for txn in data.filter_txns(entries):
            Y = vectorizer.transform([txn.narration])
            scores = classifier.decision_function(Y)
            index = scores.argmax(axis=1)
            predicted = classifier.classes_[index][0]
            score = scores[0, index].squeeze()
            if score > 0.5:
                txn.postings.append(data.Posting(predicted, None, None, None, None, None))
        return entries

class SmartImporter(Smart, Importer):
   pass
```
I have extracted this from some code I use that does more and I haven't tested it in isolation, but you get the gist of it, I believe.

**Comment by bratekarate** [comment]
Thanks first for the last example, I have no experience with ML so this may prove useful in the future.

What I meant with chaining and reproducibility: of course your example works and will be reproducible on any machine, and it's basically chaining the functions. What I meant is that it is easier to supply extra filters (I will only call hooks filters from now on) as a list than to use multiple Inheritance to override the default. To me this makes it more reproducible, as I don't have to think about any implementational details to apply different hooks on different machines. I can add and remove, change the order of filters at runtime. The logic to chain filters is at one place and I merely provide filters implementing a common interface.

I admit that it may be a matter of taste as well, I don't like multiple inheritance and usually don't use it (sorry, damaged by Java background). I appreciate the suggestion as it made me research more about python inheritance, but it's just not solution I want to use. If I have to add code myself anyway, I can just reimplement something like `apply_hooks` and thus avoid implementing or changing classes for every filter I want to apply to an importer. Even better yet, I could just use the beangulp hook functionality and wrap smart_importer (or rewrite it) to fit the interface, if I could somehow get this to run with fava.

Did not mean to criticize your approach, I'd just prefer to use another.

**Comment by dnicolodi** [comment]
Sure, you are free to rediscover the wheel in any shape you prefer. However, the issue you are commenting on is about adding a particular shape of wheel to the project, and I have been explaining why in my opinion there is not really any need for funny shaped wheels.

Linear multiple inheritance is just syntactic sugar for repeated single inheritance and monkey patching class instances as `smart_importer` does is bad form most of the time. The approach I presented can be easily extended to call what you call a list of hook functions.

**Comment by bratekarate** [comment]
The discussion about the incorporation of `apply_hooks` is long done, as I said I understand that it will not be incorporated and will not bother to change anything for myself, possibly being able to leverage the `hooks` mechanism of beangulp/beancount in the future.

Reinventing the wheel is not my intention, as it has been reinvented in another project. The thing is, I'm arguing about the interface, you are arguing about implementation. I want an actual python list that I populate with functions/filters/hooks, like `[filter1,  filter2, filter3]`. Not really clear to me why this is "what I call a list of hook functions"; it simply is an actual list of hook functions. Smart importer provides that, albeit apparently not in the best way. I will still keep using this approach until another one satisfies my, even if I understand your point. 

Maybe your approach can even be extended to provide an API for configuration the way I want it and I simply don't grasp it; of course that's possible as well. As far as I understand your example though, it does not lead to the result I want.

I never argued that monkey patching the class instance is a great way to go about things, It just happens that the resulting API is just what I want to have and so far I saw no alternatives. Thanks for taking the time discussing this anyway.

**Comment by dnicolodi** [comment]
This is turning into a condensed course on Pthon programming:
```python
class Hooked:
    def __init__(self, *args, hooks=None, **kwargs):
        super().__init__(*args, **kwargs)
        self.hooks = hooks or []

    def extract(self, file, existing_entries):
        entries = super().extract(file, existing_entries)
        for func in self.hooks:
            entries = func(entries, existing_entries)
        return entries

class ImporterWithHooks(Hooked, Importer):
   pass

impoter = ImporterWithHooks(..., hooks=[whatever, function, you, want])
```
typed directly into the Github issue comment and obviously untested.

**Comment by bratekarate** [comment]
This example is actually similar to what I thought you meant. It's is almost what I want, but it still ignores that I don't want to define one class for each Importer I want to use. I want to put all boilerplate into a package and then *only* configure the hooks for *any* importer. Obviously I want the same flexibility that the `apply_hook` decorator provides. It may be cleaner code, but still lacks the features I need.

I did some digging around myself and the `type` function seems promising to do this at runtime. Something like this:

```python
def hook_importer(importer, hooks):
    HookedImporter = type("HookedImporter", (Hooked, type(importer), {})
    hookedImporter = ... # instantiate with props of `importer` and `hooks`
    return hookedImporter
```
The only thing I haven't figured out yet is how to instantiate the new class with all the members of the provided `importer`. 

Thanks for all the ideas, I surely have enough to play around with for now. :)

**Edit**: This satisfies me:
```python
def hookify_importer(ImporterType):
    return type(f'Hooked{ImporterType.__name__}', (Hooked, ImporterType, {})


hookify_importer(MyImporter)(..., hooks=[my, hooks])
```

Maybe not pythonic, who knows :P

**Comment by dnicolodi** [comment]
Congratulations, you found the syntax to define a class without using the `class` keyword.

**Comment by bratekarate** [comment]
I'd say I found the syntax to dynamically define classes at runtime, which allows using the approach you suggested for my use case. But if you think that the absence of the class keyword is the only difference, that's fine for me too :)
